File: Visual Studio Product Team/Official Visual Studio 2010 Samples for C# 4.0/[VB]-Official Visual Studio 2010 Samples for C# 4.0/ADO.NET Entity Framework/VB/EmployeeTracker/ViewModel/EmployeeWorkspaceViewModel.vb
' Copyright © Microsoft Corporation.  All Rights Reserved.
' This code released under the terms of the 
' Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.ObjectModel
Imports System.Windows.Input
Imports EmployeeTracker.Common
Imports EmployeeTracker.Model
Imports EmployeeTracker.ViewModel.Helpers

Namespace EmployeeTracker.ViewModel

    ''' <summary>
    ''' ViewModel for managing Employees within the company
    ''' </summary>
    Public Class EmployeeWorkspaceViewModel
        Inherits ViewModelBase
        ''' <summary>
        ''' The employee currently selected in this workspace
        ''' </summary>
        Private privateCurrentEmployee As EmployeeViewModel

        ''' <summary>
        ''' UnitOfWork for managing changes
        ''' </summary>
        Private unitOfWork As IUnitOfWork

        ''' <summary>
        ''' Departments to be used for lookups
        ''' </summary>
        Private departmentLookup As ObservableCollection(Of DepartmentViewModel)

        ''' <summary>
        ''' Initializes a new instance of the EmployeeWorkspaceViewModel class.
        ''' </summary>
        ''' <param name="employees">Employees to be managed</param>
        ''' <param name="departmentLookup">The departments to be used for lookups</param>
        ''' <param name="unitOfWork">UnitOfWork for managing changes</param>
        Public Sub New(ByVal employees As ObservableCollection(Of EmployeeViewModel), ByVal departmentLookup As ObservableCollection(Of DepartmentViewModel), ByVal unitOfWork As IUnitOfWork)
            If employees Is Nothing Then
                Throw New ArgumentNullException("employees")
            End If

            If departmentLookup Is Nothing Then
                Throw New ArgumentNullException("departmentLookup")
            End If

            If unitOfWork Is Nothing Then
                Throw New ArgumentNullException("unitOfWork")
            End If

            Me.unitOfWork = unitOfWork
            Me.AllEmployees = employees
            Me.departmentLookup = departmentLookup
            Me.CurrentEmployee = If(employees.Count > 0, employees(0), Nothing)

            ' Re-act to any changes from outside this ViewModel
            AddHandler Me.AllEmployees.CollectionChanged, Sub(sender, e)
                                                              If e.OldItems IsNot Nothing AndAlso e.OldItems.Contains(Me.CurrentEmployee) Then
                                                                  Me.CurrentEmployee = Nothing
                                                              End If
                                                          End Sub

            Me.AddEmployeeCommand = New DelegateCommand(Sub(o) Me.AddEmployee())
            Me.DeleteEmployeeCommand = New DelegateCommand(Sub(o) Me.DeleteCurrentEmployee(), Function(o) Me.CurrentEmployee IsNot Nothing)
        End Sub

        ''' <summary>
        ''' Gets the command for adding a new employee
        ''' </summary>
        Private privateAddEmployeeCommand As ICommand
        Public Property AddEmployeeCommand() As ICommand
            Get
                Return privateAddEmployeeCommand
            End Get
            Private Set(ByVal value As ICommand)
                privateAddEmployeeCommand = value
            End Set
        End Property

        ''' <summary>
        ''' Gets the command for deleting the current employee
        ''' </summary>
        Private privateDeleteEmployeeCommand As ICommand
        Public Property DeleteEmployeeCommand() As ICommand
            Get
                Return privateDeleteEmployeeCommand
            End Get
            Private Set(ByVal value As ICommand)
                privateDeleteEmployeeCommand = value
            End Set
        End Property

        ''' <summary>
        ''' Gets all employees whithin the company
        ''' </summary>
        Private privateAllEmployees As ObservableCollection(Of EmployeeViewModel)
        Public Property AllEmployees() As ObservableCollection(Of EmployeeViewModel)
            Get
                Return privateAllEmployees
            End Get
            Private Set(ByVal value As ObservableCollection(Of EmployeeViewModel))
                privateAllEmployees = value
            End Set
        End Property

        ''' <summary>
        ''' Gets or sets the employee currently selected in this workspace
        ''' </summary>
        Public Property CurrentEmployee() As EmployeeViewModel
            Get
                Return Me.privateCurrentEmployee
            End Get

            Set(ByVal value As EmployeeViewModel)
                Me.privateCurrentEmployee = value
                Me.OnPropertyChanged("CurrentEmployee")
            End Set
        End Property

        ''' <summary>
        ''' Handles addition a new employee to the workspace and model
        ''' </summary>
        Private Sub AddEmployee()
            Dim emp As Employee = Me.unitOfWork.CreateObject(Of Employee)()
            Me.unitOfWork.AddEmployee(emp)

            Dim vm As New EmployeeViewModel(emp, Me.AllEmployees, Me.departmentLookup, Me.unitOfWork)
            Me.AllEmployees.Add(vm)
            Me.CurrentEmployee = vm
        End Sub

        ''' <summary>
        ''' Handles deletion of the current employee
        ''' </summary>
        Private Sub DeleteCurrentEmployee()
            Me.unitOfWork.RemoveEmployee(Me.CurrentEmployee.Model)
            Me.AllEmployees.Remove(Me.CurrentEmployee)
            Me.CurrentEmployee = Nothing
        End Sub
    End Class
End Namespace

File: Visual Studio Product Team/Official Visual Studio 2010 Samples for C# 4.0/46872-Official Visual Studio 2010 Samples for C# 4.0/ADO.NET Entity Framework/C#/EmployeeTracker/ViewModel/EmployeeWorkspaceViewModel.cs
// Copyright © Microsoft Corporation.  All Rights Reserved.
// This code released under the terms of the 
// Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)


namespace EmployeeTracker.ViewModel
{
    using System;
    using System.Collections.ObjectModel;
    using System.Windows.Input;
    using EmployeeTracker.Common;
    using EmployeeTracker.Model;
    using EmployeeTracker.ViewModel.Helpers;

    /// <summary>
    /// ViewModel for managing Employees within the company
    /// </summary>
    public class EmployeeWorkspaceViewModel : ViewModelBase
    {
        /// <summary>
        /// The employee currently selected in this workspace
        /// </summary>
        private EmployeeViewModel currentEmployee;

        /// <summary>
        /// UnitOfWork for managing changes
        /// </summary>
        private IUnitOfWork unitOfWork;

        /// <summary>
        /// Departments to be used for lookups
        /// </summary>
        private ObservableCollection<DepartmentViewModel> departmentLookup;

        /// <summary>
        /// Initializes a new instance of the EmployeeWorkspaceViewModel class.
        /// </summary>
        /// <param name="employees">Employees to be managed</param>
        /// <param name="departmentLookup">The departments to be used for lookups</param>
        /// <param name="unitOfWork">UnitOfWork for managing changes</param>
        public EmployeeWorkspaceViewModel(ObservableCollection<EmployeeViewModel> employees, ObservableCollection<DepartmentViewModel> departmentLookup, IUnitOfWork unitOfWork)
        {
            if (employees == null)
            {
                throw new ArgumentNullException("employees");
            }

            if (departmentLookup == null)
            {
                throw new ArgumentNullException("departmentLookup");
            }

            if (unitOfWork == null)
            {
                throw new ArgumentNullException("unitOfWork");
            }

            this.unitOfWork = unitOfWork;
            this.AllEmployees = employees;
            this.departmentLookup = departmentLookup;
            this.CurrentEmployee = employees.Count > 0 ? employees[0] : null;

            // Re-act to any changes from outside this ViewModel
            this.AllEmployees.CollectionChanged += (sender, e) =>
            {
                if (e.OldItems != null && e.OldItems.Contains(this.CurrentEmployee))
                {
                    this.CurrentEmployee = null;
                }
            };

            this.AddEmployeeCommand = new DelegateCommand((o) => this.AddEmployee());
            this.DeleteEmployeeCommand = new DelegateCommand((o) => this.DeleteCurrentEmployee(), (o) => this.CurrentEmployee != null);
        }

        /// <summary>
        /// Gets the command for adding a new employee
        /// </summary>
        public ICommand AddEmployeeCommand { get; private set; }

        /// <summary>
        /// Gets the command for deleting the current employee
        /// </summary>
        public ICommand DeleteEmployeeCommand { get; private set; }

        /// <summary>
        /// Gets all employees whithin the company
        /// </summary>
        public ObservableCollection<EmployeeViewModel> AllEmployees { get; private set; }

        /// <summary>
        /// Gets or sets the employee currently selected in this workspace
        /// </summary>
        public EmployeeViewModel CurrentEmployee
        {
            get
            {
                return this.currentEmployee;
            }

            set
            {
                this.currentEmployee = value;
                this.OnPropertyChanged("CurrentEmployee");
            }
        }

        /// <summary>
        /// Handles addition a new employee to the workspace and model
        /// </summary>
        private void AddEmployee()
        {
            Employee emp = this.unitOfWork.CreateObject<Employee>();
            this.unitOfWork.AddEmployee(emp);

            EmployeeViewModel vm = new EmployeeViewModel(emp, this.AllEmployees, this.departmentLookup, this.unitOfWork);
            this.AllEmployees.Add(vm);
            this.CurrentEmployee = vm;
        }

        /// <summary>
        /// Handles deletion of the current employee
        /// </summary>
        private void DeleteCurrentEmployee()
        {
            this.unitOfWork.RemoveEmployee(this.CurrentEmployee.Model);
            this.AllEmployees.Remove(this.CurrentEmployee);
            this.CurrentEmployee = null;
        }
    }
}
